### ISSUE TITLE
HeteroFL

### ISSUE THREAD

**Issue by jafermarq** [adap/flower#2037]
# HeteroFL

* **Title:** HeteroFL: Computation and Communication Efficient Federated Learning for Heterogeneous Clients
* **Venue:** ICLR 2021
* **Link to paper:** https://openreview.net/forum?id=TNkPBBYFkXg

## Do you want to work on this baseline? 

> 🌻 Check everything about the `Summer of Reproducibility` on [`flower.dev/summer`](https://flower.dev/summer/)

> All available baselines are listed in the [Summer of Reproducibility Dashboard](https://github.com/orgs/adap/projects/6/) and also in the [GitHub Issues with the `summer-of-reproducibility` label](https://github.com/adap/flower/labels/summer-of-reproducibility). The content is the same.

> 📝 It is advised to complete these steps before your start working on your code. But if you can't wait to implement your baseline with Flower (we totally understand it 😄), please ensure you follow the steps on [how to contribute a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline).

What follows are the steps 1 & 2 in the Summer of Reproducibility instructions.

### 1. Join the Summer of Reproducibility program
- [x] Join the [Flower Slack](https://flower.dev/join-slack/) and say _"hi! 👋"_ in the channel `#summer-of-reproducibility`.
- [x] Pick a baseline from our curated list **<---------------------------------------- [you are doing this now]**
### 2. Define the scope of your contribution
- [x] **What are you going to reproduce?** Add a comment to your issue and tell us about your plan regarding this baseline: what experiments from the paper are you reproducing?, for which datasets ? the more details you provide us with the better !
- [x] **Check if you are eligible for a reward**. 
    > As we have to comply with US/EU regulations, we have checked that individual contributors based on these countries or territories are eligible: Australia, Austria, Belgium, Bulgaria, Canada, Croatia, Cyprus, Czech Republic, Denmark, Estonia, Finland, France, Germany, Gibraltar, Greece, Hong Kong SAR China, Hungary, India, Ireland, Italy, Japan, Latvia, Liechtenstein, Lithuania, Luxembourg, Malta, Mexico, Netherlands, New Zealand, Norway, Poland, Portugal, Romania, Singapore, Slovakia, Slovenia, South Korea, Spain, Sweden, Switzerland, Thailand, United Arab emirates, United Kingdom United States.

    If where you are based is not on the list, please send us an email (`summer@flower.dev`) letting us know a bit about yourself (where are you currently based?, are you a university student? do you work at a public institution?). Please tell us the baselines you are interested in implementing (i.e. tell us your GitHub issue if you have crated one). We will reach back to you.
- [x] We will discuss with you about your contribution plan, if it sounds like a substantial enough contribution according to the Summer of Reproducibility rules (check our website [flower.dev/summer](https://flower.dev/summer/)), you'll get the OK from us to start working on your baseline!

## What happens next?

- [x] This item will be moved to the `In Progress` stage by a member of the Flower Team.

- [x] Follow the [instructions for creating a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline) which will guide you through the process step-by-step.

## Is something wrong or not clear ?
* Ask a question directly in your issue.
* Reach out to us via the [Flower Slack](https://flower.dev/join-slack/) and ask your question in the #summer-of-reproducibility channel
* Check all the details (including FAQ) in the Summer of Reproducibility website: flower.dev/summer

**Comment by msck72** [comment]
Hi @jafermarq,
I would like to implement HeteroFL baseline,

**Contribution plan:**
- Implementing the HeteroFL Strategy in Flower framework using pytorch
- Reproduce the results in figure 2 for fixed computational complexities and table 2 for dynamic computational complexities using the CIFAR10 dataset (data partition being same as in McMahan FedAvg paper) and compare the results with FedAvg
- Further, implementing the Table 1 using MNIST dataset (dynamic computational complexities)

I confirm that I belong to country eligible for reward

**Comment by jafermarq** [comment]
Hi @msck72, `HeteroFL` + `PyTorch` + `Flower` sounds like a perfect mix! So the plan is to reproduce Figure 2 and Table 1&2 including all the model configurations and FedAvg? How about the non-IID setups ?

Thanks for confirming your eligibility.

**Comment by msck72** [comment]
Hey, Yes, I would like to reproduce Figure 2 and table 1 & 2 (all model configurations) and FedAvg, including the non-IID setups.

**Comment by msck72** [comment]
Hey, Yes, I would like to reproduce Figure 2 on (IID CIFAR10) for fixed computational complexities and table1&2 which also includes non-IID on CIFAR10 and MNIST for dynamic computational complexities(all model configurations) and FedAvg. 

And, shall I also include Figure 5(non IID MNIST) and 6 (non IID Cifar10) for fixed computational complexities

**Comment by jafermarq** [comment]
Ok, thanks for confirming! The you are all set to start participating in the Flower _Summer of Reproducibility_. Remember we are almost one month in, so there is just 2 months left. We encourage all contributors to start as soon as they can.

I have now ✅ all the points in Step 1&2 above, added you as the _assignee_ to this GitHub issue, and I have moved this baseline to the _In Progress_ stage. You can find additional details on how to begin with all things code related by following the link in the _What happens next?_ section at the end of the issue description. Feel free to ask questions on our Slack workspace if you encounter some problems or leave as a comment also if you have some suggestions on how to improve things. 

Looking forward to seeing your `HeteroFL` baseline in action! 🚀

**Comment by negedng** [comment]
Hi,

I have an implementation of HeteroFL (ml-framework agnostic), at least the model-agnostic part. I haven't done a full comparison with the original code, but it learns and produces results reasonably similar to the expected one. I am looking forward to your results!

Here are snippets of my code, in case it helps. All in numpy, and not assuming channel-first or channel-last so it can be used with both TF and PyTorch:
1) Big model to small model:
```
def take(a, new_shape):
    """Takes the top-left submatrix with given shape"""
    a_shape = np.shape(a)
    z = a
    for i, d in enumerate(new_shape):
        z_shape = np.shape(z)
        if z_shape[i] != d:
            take_index_list = range(d)
    z = z.take(take_index_list, axis=i)
    return z
```
You can pass the parameters from the client.fit() for the small clients into this layer-by-layer to get the top-left submatrix.

2) Aggregation
```
def aggregate_hetero(results):
    """Compute weighted average with different model sizes."""

    def aggregate_layer(layer_updates, num_examples_list):
        """Padding layers with 0 to max size, then average them"""
        # Get the layer's largest form
        max_ch = np.max([np.shape(l) for l in layer_updates], axis=0)
        layer_agg = np.zeros(max_ch)
        count_layer = np.zeros(max_ch)  # to average by num of models that size
        for l, num in zip(layer_updates, num_examples_list):
            local_ch = np.shape(l)
            pad_shape = [(0, a) for a in (max_ch - local_ch)]
            l_padded = np.pad(l, pad_shape, constant_values=0.0)
            ones_of_shape = np.ones(local_ch) * num
            ones_pad = np.pad(ones_of_shape, pad_shape, constant_values=0.0)
            count_layer = np.add(count_layer, ones_pad)
            layer_agg = np.add(layer_agg, l_padded)
        if np.any(count_layer == 0.0):
            print(count_layer)
            raise ValueError("Diving with 0")
        layer_agg = layer_agg / count_layer
        return layer_agg

    # Calculate the total number of examples used during training
    num_examples_list = [num_examples for _, num_examples in results]

    # Create a list of weights, each multiplied by the related number of examples
    weighted_weights = [
        [layer * num_examples for layer in weights] for weights, num_examples in results
    ]
    # Aggregate the layers
    agg_layers = [aggregate_layer(l, num_examples_list) for l in zip(*weighted_weights)]
    return agg_layers
```
You just have to overwrite the FedAvg's aggregate function with this one (it follows the same structure).
I have some unit-tests to check if it does the same as HeteroFL, but that's not easily portable

It's totally up to you if you use my code or not, I would've shared it sooner or later anyways when my current research goes to publication. :)

**Comment by msck72** [comment]
Hi @negedng,
Thank you for the suggestion and code snippets, will sure consider them while implementing.

**Comment by jafermarq** [comment]
Congratulations @msck72 your baseline has been merged. We'll follow up with you soon to complete your _Summer of Reproducibility_ journey 😄 ! Merry Christmas!
